### ISSUE TITLE
How can I help PureScript to generate faster code?

### ISSUE THREAD

**Issue by rvion** [purescript/purescript#2267]
is there any work we can do to help speed up purescript generated code ?

It slowly turns out that the currently generated code is too slow for most projects I was planning to do with purescript 😢 😢 

so many little things are making it slow:
 - "replicateA 10.000 doSomethingFast" is taking way too long because of TCO not properly running
 - saturated function call are not uncurried
 - no local optimization pass
 - etc.

I saw some issues and attemps to improve the situation, but I can't find some good roadmap or concrete action plan. 

Is there such a roadmap ?
How can I help ?

**Comment by texastoland** [comment]
#1976 is my biggest wish.

**Comment by paf31** [comment]
That and some other optimizations are handled more uniformly by inlining across module boundaries. You might want to talk to @rightfold about https://github.com/iron-codegen/iron.

**Comment by jutaro** [comment]
I am very interested in enhaced purescript performance (webgl animations...). I started with the uncurry optimization, which is a javascript AST transformation. See [Optimizing Purescript's Runtime Performance](http://symbolian.net/2016/04/07/513/) for a description. On this level two more optimizations are obvious (copied from the article):

* Inlining: Assign a weight to a function definition based on the size of its body and the number of call sites it's referenced from; inline the code when it's sufficiently small or gets called from just a single (or very few) site(s).

* An equivalent of Lambda Calculus' eta conversion: Replace a function f(x,..) = g(x,..) with g(x,..)

When typeclasses gets specialized at code generation time we will modify the uncurry optimization so that it will be applied too. 

By the way, I commented about the bad performance of replicate as well, which is now about 20 times slower, then the implementation in Array was before! See [Recover replicate](https://github.com/purescript/purescript-arrays/issues/67 ). The opinion that users of purescript don't need good performance (as stated in the thread) is for me not valid, and a replicate function which is 20 times slower may be abstract and beautiful, but is unusable in practice, and I hope purescript will not progress further in this direction.

**Comment by garyb** [comment]
#687 is something that will help with performance a fair bit, especially in the case of code that relies on typeclasses.

I have quite a few ideas for optimisations and want to experiment with supercompilation but just haven't had the time to do so.

Part of the reason the compiler currently does almost no optimisation is that it can make codegen unpredictable, something that Phil has stated several times he'd like to avoid in the main compiler. I can't say I'm bothered either way about that personally, as the only time I look at the generated JS is when debugging the compiler, plus we have source maps, but I understand the motivation. So the chances are a backend that optimises more would be a separate thing, or rolled into `psc-publish` so that whole program optimisations can be applied (like uncurrying), etc.

> The opinion that users of purescript don't need good performance (as stated in the thread)

I'm not really sure why you still use PureScript given that you're so hostile towards our views about improving the language. I'm glad that you still do use it, but it seems that you're unhappy with something in every release we make, and you choose to misinterpret our position on performance every time it is discussed.

This is the very last time I'll engage in this kind of discussion with you, so let me state my thoughts once and for all so I can link back to this comment the next time it arises:

1. Performance is not unimportant, but nor is it king. A big part of the reason PureScript exists is to enable use of abstractions that make code more flexible and improve correctness. That sometimes comes at the expense of performance, but the desire for correctness and expressiveness is why we're not just writing JavaScript.
2. There are only a certain number of hours in the day. Nobody is getting paid to work on the compiler (directly). The language needs work in several areas. With those factors in mind, when I get the time to work on things, I prefer to work on things that are broken/buggy/missing. In the not too distant future (I hope) things will be good enough in the areas that I can help with and I'll turn my attention to the aforemetioned supercompiler/optimisation passes.
3. The core libraries are not the language. You do not have to use them. If the philosophy of the general purpose libraries does not align with your own or is unsuitable for your application you can fork them and/or write your own specialised versions that better suit your needs. This is part of the language philosophy and why we don't include our `Prelude` as standard, etc.

In summary, I would love for PureScript to be as fast as it possibly could be, but not at the expensive of avoiding the things that make the language great to use in the first place. I would also love to work on a much better optimising backend (apart from anything else, I think it would be fun), but I do not have the time to work on it yet. Since I'm selfish and mainly work on things that affect me personally, the current performance is not painful enough to pull me away from working on other things that are imperfect or broken.

**Comment by paf31** [comment]
Closing. Please consult the issue tracker, specifically the `optimizer` tag.

Now that we have `--dump-corefn`, it should also be possible to implement specific optimizations on the functional core in external tools.
